flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -329,37 +332,47 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
         // materialization may truncate only a part of the previous result and the backend would
         // have to split it somehow for the former option, so the latter is used.
         lastCheckpointId = checkpointId;
-        lastUploadedFrom = materializedTo;
+        lastUploadedFrom = periodicMaterializer.getMaterializedState().lastMaterializedTo();
         lastUploadedTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
 
         LOG.debug(
                 "snapshot for checkpoint {}, change range: {}..{}",
                 checkpointId,
                 lastUploadedFrom,
                 lastUploadedTo);
+
+        MaterializedState materializedStateCopy = periodicMaterializer.getMaterializedState();
+
         return toRunnableFuture(
                 stateChangelogWriter
                         .persist(lastUploadedFrom)
-                        .thenApply(this::buildSnapshotResult));
+                        .thenApply(delta -> buildSnapshotResult(delta, materializedStateCopy)));
     }
 
-    private SnapshotResult<KeyedStateHandle> buildSnapshotResult(ChangelogStateHandle delta) {
-        // Can be called by either task thread during the sync checkpoint phase (if persist future
-        // was already completed); or by the writer thread otherwise. So need to synchronize.
-        // todo: revisit after FLINK-21357 - use mailbox action?
-        synchronized (materialized) {
-            // collections don't change once started and handles are immutable
-            List<ChangelogStateHandle> prevDeltaCopy = new ArrayList<>(restoredNonMaterialized);
-            if (delta != null && delta.getStateSize() > 0) {
-                prevDeltaCopy.add(delta);
-            }
-            if (prevDeltaCopy.isEmpty() && materialized.isEmpty()) {
-                return SnapshotResult.empty();
-            } else {
-                return SnapshotResult.of(
-                        new ChangelogStateBackendHandleImpl(
-                                materialized, prevDeltaCopy, getKeyGroupRange()));
-            }
+    @Override
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        periodicMaterializer.triggerMaterialization();
+    }
```

Comment:
This is very..practical :slightly_smiling_face: 
I think a similar effect can be achieved by using some custom `ScheduledExecutorService` in Materializer (e.g. `ManuallyTriggeredScheduledExecutorService`)

Reply:
:-)

Reply:
I've played with `ManuallyTriggeredScheduledExecutorService` a bit
It does not make things cleaner or less exposure of methods/APIs, but introducing more complexities for PeriodicMaterializer constructor (now I have to expose a way to substitute the executor).

The reason here is I want to test materialization itself works correctly. To do that, I have to expose the "triggerMaterialization" method, and trigger it through the `ManuallyTriggeredScheduledExecutorService` manually.

so "triggerMaterialization" has to be exposed anyways. 

Please let me knwo whether this makes sense to you, or if you have better suggestions.

Reply:
I tried to disable materialization by commenting it in `ChangelogKeyedStateBackend.triggerMaterialization` and the test still passed, so I think just triggering is not enough.

This suggests, that if we want to test it on this level, a more fine-grained control over materialization is needed; which can be gained by a direct access to the materializer in test.

And if we have such access, exposing something through the backend isn't necessary.
But probably it makes sense to first resolve the questions about backend-materializer interaction (like [this](https://github.com/apache/flink/pull/16606#discussion_r682621905) one).

WDYT?

Reply:
> I tried to disable materialization by commenting it in `ChangelogKeyedStateBackend.triggerMaterialization` and the test still passed, so I think just triggering is not enough.

That's expected, that's equivalent to disable materialization that uses purely changelog.

Reply:
As we discussed offline, this only works because a direct executor is used when creating materilizier (by "works" I mean the call waits for the materialization).
Please add a comment at the very least.

However, I still think the test or production code structure is wrong. With a proper separation between materializer and backend, we should be able to listen for materialization completion and take a snapshot only after that (instead of relying on the internal materializer behavior).